File: sorting/problem56/problem56.py
from typing import List


class Solution:
    def merge(self, intervals: List[List[int]]) -> List[List[int]]:
        # Sort by start; extend or append current interval
        intervals.sort(key=lambda x: x[0])
        res: List[List[int]] = [intervals[0][:]]

        for start, end in intervals[1:]:
            if start <= res[-1][1]:
                res[-1][1] = max(res[-1][1], end)
            else:
                res.append([start, end])
        return res

File: sorting/problem56/problem56.cpp
#include <algorithm>
#include <vector>

using namespace std;

class Solution {
public:
    vector<vector<int>> merge(vector<vector<int>>& intervals) {
        sort(intervals.begin(), intervals.end());
        vector<vector<int>> res;
        for (const auto& inv : intervals) {
            if (res.empty() || inv[0] > res.back()[1]) {
                res.push_back(inv);
            } else {
                res.back()[1] = max(res.back()[1], inv[1]);
            }
        }
        return res;
    }
};
